File: designpatterns/java/src/command/RemoteControl.java
package command;

public class RemoteControl {
    Command[] onCommands;
    Command[] offCommands;

    public RemoteControl() {
        onCommands = new Command[7];
        offCommands = new Command[7];

        Command noCommand = new NoCommand();
        for(int i = 0; i < 7; i++) {
            onCommands[i] = noCommand;
            offCommands[i] = noCommand;
        }
    }

    public void setCommand(int slot, Command onCommand, Command offCommand) {
        onCommands[slot] = onCommand;
        offCommands[slot] = offCommand;
    }

    public void onButtonWasPushed(int slot) {
        onCommands[slot].execute();
    }

    public void offButtonWasPressed(int slot) {
        offCommands[slot].execute();
    }

    public String toString() {
        StringBuffer stringBuffer = new StringBuffer();
        stringBuffer.append("\n------ Remote Control ------\n");
        for(int i = 0; i < onCommands.length; i++) {
            stringBuffer.append("[slot ").append(i).append("] ")
                    .append(onCommands[i].getClass().getName())
                    .append("    ").append(offCommands[i].getClass().getName()).append("\n");
        }
        return stringBuffer.toString();
    }

}

File: designpatterns/kotlin/designpatterns/src/main/kotlin/command/example/RemoteControl.kt
package command.example

data class RemoteControl(
    private val slotTotal: Int,
    private val onCommand: List<Command>,
    private val offCommand: List<Command>,
    private var undoCommand: Command? = null,
) {

    init {
        if (onCommand.size > slotTotal || offCommand.size > slotTotal) {
            throw RuntimeException("슬롯 할당 갯수를 초과했습니다.")
        }
    }

    fun onButtonWasPressed(slot: Int) {
        onCommand[slot].execute()
        undoCommand = onCommand[slot]
    }

    fun offButtonWasPressed(slot: Int) {
        offCommand[slot].execute()
        undoCommand = offCommand[slot]
    }

    fun undoButtonWasPushed() {
        undoCommand?.undo()
    }

}